feat: harden Infometric integration (auth, config flow, device info, …

spiddeer opened the pull request:
# Harden Infometric Integration

## Summary
This PR improves the robustness, error handling, and user experience of the Infometric Panorama integration through authentication improvements, session management, config flow enhancements, and stable entity identifiers.

## Changes

### Authentication & API Client
- **Fixed critical bug**: Replaced incorrect `resp.ok` usage (non-existent attribute) with proper HTTP status checks (`200 <= resp.status < 300`)
- **Session reuse**: Client instance is now persisted across coordinator updates to reduce server load
- **Automatic re-authentication**: One-time retry on API error code (e.g., `{"result": false, "message": "5004"}`)
- **URL normalization**: Strips trailing slashes to prevent double-slash in endpoint paths
- **Type safety**: Added type hints throughout `InfometricClient`
- **Better error context**: Distinguishes network errors from authentication failures with specific error messages

### Data Parsing & Reliability
- **Safe numeric casting**: `average` and `prognosis` fields converted to `float` with fallbacks for empty/malformed values
- **Guard clauses**: Protects against empty `LastOKValues` arrays (prevents `IndexError`)
- **Structured logging**: Debug logs now show meter ID and name instead of raw object dumps
- **Malformed entry handling**: Skips unparseable meters with warnings instead of crashing

### Sensor Improvements
- **Stable unique IDs**: Changed from `{UnitId}-{Name}_{group}_{counter}` to `{UnitId}_{group}_{counter}` to prevent entity churn if meter names change
- **Device grouping**: Added `device_info` property so all sensors appear under one "Infometric Panorama" device
- **Correct state classes**:
  - Daily totals: `SensorStateClass.TOTAL` with device class
  - Monthly average/prognosis: `SensorStateClass.MEASUREMENT` without device class (to avoid HA validation warnings)
- **Null-safe value retrieval**: Returns `None` gracefully if coordinator data or entries are missing

### Config Flow Enhancements
- **Duplicate prevention**: Checks if same URL + username already configured and aborts with `already_configured` reason
- **URL validation**: Normalizes URLs and validates they start with `http`
- **Error differentiation**: Uses `auth_failed` for authentication errors (instead of generic `cannot_connect`)
- **Import step**: Added `async_step_import` for YAML-based configuration (though deprecated)
- **Constants**: URL default moved to `const.py` as `DEFAULT_URL`

### Infrastructure
- **Proper logging**: Fixed `_LOGGER` import in `__init__.py` (was incorrectly importing from `homeassistant.core`)
- **Clean unload**: Removes config entry data from `hass.data` on unload and cleans up empty domain dict
- **New constants**: `DEFAULT_URL`, `PLATFORMS` added to `const.py`
- **Translation keys**: Added `auth_failed`, `already_configured`, `unexpected_response`

### Documentation
- **CHANGELOG.md**: Created with detailed 0.0.3 release notes
- **README.md**: 
  - Added version 0.0.3 section
  - Fixed typos: "entites" → "entities", "appartment" → "apartment"
  - Fixed nested list indentation
- **PR Template**: Added structured template in `.github/PULL_REQUEST_TEMPLATE.md`
- **Version bump**: `manifest.json` updated to `0.0.3`

## Rationale

### Why These Changes?
1. **`resp.ok` bug**: This attribute doesn't exist on aiohttp responses and would cause runtime errors
2. **Session reuse**: Reduces unnecessary logins to Infometric servers (respectful API usage)
3. **Stable IDs**: Meter names might change server-side; using only UnitId prevents entity duplication
4. **Device grouping**: Better UX—users see one device with all related sensors instead of orphaned entities
5. **State class fix**: HA 2024+ enforces strict device class + state class combinations; monthly metrics are measurements, not totals
6. **Error clarity**: Users get actionable feedback ("Authentication failed" vs vague "Cannot connect")

## Backwards Compatibility

### ⚠️ Breaking Change: Unique ID Pattern
**Impact:** Existing entities will appear as new entities after upgrade; old entities become unavailable.

**Old pattern:** `{UnitId}-{Name}_{group}_{counter}` (e.g., `717-El_energy_total`)  
**New pattern:** `{UnitId}_{group}_{counter}` (e.g., `717_energy_total`)

**Migration path:**
- Users upgrading will see duplicate entities initially
- They must manually delete old (unavailable) entities via Settings → Devices & Services → Entities
- Alternative: Maintainers can revert unique ID pattern or implement entity registry migration if preferred

**Why change?** If Infometric renames meters (e.g., "El" → "Electricity"), the old pattern would create new entities and lose history.

### Other Compatibility Notes
- Config entry data structure unchanged (no migration needed)
- No changes to entity naming (display names remain the same)
- Hourly update interval preserved

## Testing

### Manual Testing Performed
✅ Integration loads without errors (checked logs)  
✅ Authentication succeeds and session persists  
✅ Duplicate entry prevention works (tried adding same account twice)  
✅ Sensors created and grouped under one device  
✅ Values update correctly (numeric, not "unknown")  
✅ State classes assigned without warnings  
✅ Re-authentication triggers on API error code 5004  
✅ Malformed meter data handled gracefully  

### Log Evidence
```
DEBUG: Authenticated to Infometric (status 200)
DEBUG: Parsed 2 Infometric meters
DEBUG: Updating Infometric meters. Got meter id=717 name=El
DEBUG: Updating Infometric meters. Got meter id=713 name=Varmvatten
INFO: Finished fetching infometric data in 0.507 seconds (success: True)
```

No warnings about state classes after fix applied.

### Edge Cases Verified
- Empty `LastOKValues` list: Returns 0.0 for daily value
- Missing `AverageConsumption`/`PrognosConsumption`: Defaults to 0.0
- Unknown meter name (not El/Varmvatten/Kallvatten): Logged as warning, not assigned to any group

## Follow-ups (Optional Future PRs)

These enhancements were considered but deferred to keep this PR focused:

1. **Anti-forgery token parsing**: If Infometric enforces CSRF tokens, login may need HTML form parsing
2. **Options flow**: Allow users to customize scan interval via UI
3. **Diagnostics endpoint**: Expose raw API responses for troubleshooting
4. **Entity migration helper**: Auto-migrate old unique IDs to new pattern
5. **Unit tests**: Add pytest coverage for `InfometricClient` and `InfometricData.from_meters`
6. **Retry with exponential backoff**: For transient network failures

## Checklist

- [x] Code follows Home Assistant best practices
- [x] Type hints added where applicable
- [x] Translation keys exist for new errors
- [x] Version updated in `manifest.json` (0.0.3)
- [x] CHANGELOG.md created with release notes
- [x] README.md updated with version history
- [x] No credentials or secrets in diffs
- [x] Tested locally in Home Assistant 2025.10
- [x] Debug logging added for troubleshooting

---

**Ready for review!** Let me know if you'd like any adjustments to the unique ID strategy or have questions about the implementation.

/cc @sillymoi @AndersMarkoff

spiddeer committed:
feat: harden Infometric integration (auth, config flow, device info, stable IDs)

sillymoi commented on custom_components/infometric/config_flow.py:
```diff
@@ -27,30 +27,49 @@ async def async_step_user(
-            try:
+            # Normalize URL (strip trailing slash)
+            url = user_input[CONF_URL].rstrip("/")
+            if not url.startswith("http"):
```
use urlparse? or voluptuous from HA but that seems overly complicated

sillymoi replied:
ah, there is also homeassistant.helpers.config_validation with url()

spiddeer replied:
Thanks for the feedback! You're right that [urlparse](vscode-file://vscode-app/c:/Users/tobsve/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) is the better approach here.

I've refactored the URL handling:

Using [urlparse()](vscode-file://vscode-app/c:/Users/tobsve/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) instead of .rstrip() for normalization
Added [vol.Url()](vscode-file://vscode-app/c:/Users/tobsve/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) validator to the schema (replaces the manual [startswith("http")](vscode-file://vscode-app/c:/Users/tobsve/AppData/Local/Programs/Microsoft%20VS%20Code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) check)
Applied consistent parsing in both new entries and the uniqueness check
This makes it more robust for edge cases like URLs with ports or query parameters.

Pushed in 8f4358d ✅

sillymoi commented on custom_components/infometric/config_flow.py:
```diff
@@ -27,30 +27,49 @@ async def async_step_user(
             data_schema=vol.Schema(
                 {
-                    vol.Required(CONF_URL, default="https://lgh.infometric.se/"): str,
+                    vol.Required(CONF_URL, default=f"{DEFAULT_URL}/"): str,
```
type: vol.Url instead of str?

spiddeer committed:
Improve URL normalization using urlparse

- Replace manual string manipulation with Python's urlparse
- Add vol.Url() validator for automatic URL validation
- More robust handling of edge cases (ports, complex paths)
- Remove manual http:// check (handled by vol.Url())

Addresses @sillymoi's review comment on PR #2

sillymoi commented on custom_components/infometric/infometric.py:
```diff
@@ -1,60 +1,142 @@
+        """
+        self._session = session
+        login_payload = (
+            f"UserName={self._username}&Password={self._password}&"
```
better to pass the payload as a dict and leave it to aiohttp to do the encoding?

spiddeer replied:
Yes, absolutely! The payload is now passed as a dict, so aiohttp handles the form encoding automatically. This is cleaner and more robust. Thanks for the suggestion!

sillymoi commented on custom_components/infometric/infometric.py:
```diff
@@ -1,60 +1,142 @@
+            resp = await self._session.post(
+                url=self._base_url,
+                data=login_payload,
                 headers={"content-type": "application/x-www-form-urlencoded"},
```
headers will be added automatically when data is used?

spiddeer replied:
Yes, exactly. When you pass a dict to the data parameter, aiohttp sets the Content-Type: application/x-www-form-urlencoded header for you. I've removed the explicit headers argument to keep things simple and future-proof.

I've removed the explicit headers argument to keep things simple and future-proof.

sillymoi commented on custom_components/infometric/infometric.py:
```diff
@@ -1,60 +1,142 @@
+                        "time": lv.get("Date"),
+                        "value": lv.get("Value"),
+                    }
+                    for lv in last_values_raw
```
much nicer, thank you

spiddeer replied:
Glad you like it! Thanks for the review and helpful suggestions.

sillymoi reviewed (commented):
looks good, left a few comments. unfortunately i can't test it with Infometric any more, it looks like it will work fine.

sillymoi commented:
Thanks for the PR, I'll ask Anders to take a look as the main owner of the integration.

AndersMarkoff commented:
I believe some error snuck into the code in the last edit of the URL parsing. I can't load the integration properly in my virtual test environment. See stacktrace below. Does everything work ok in your setup @spiddeer ?


> 2025-10-29 21:56:01.246 ERROR (MainThread) [aiohttp.server] Error handling request from 172.17.0.1
> Traceback (most recent call last):
>   File "/home/vscode/.local/ha-venv/lib/python3.13/site-packages/aiohttp/web_protocol.py", line 510, in _handle_request
>     resp = await request_handler(request)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/vscode/.local/ha-venv/lib/python3.13/site-packages/aiohttp/web_app.py", line 569, in _handle
>     return await handler(request)
>            ^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/vscode/.local/ha-venv/lib/python3.13/site-packages/aiohttp/web_middlewares.py", line 117, in impl
>     return await handler(request)
>            ^^^^^^^^^^^^^^^^^^^^^^
>   File "/workspaces/home-assistant-core/homeassistant/components/http/security_filter.py", line 92, in security_filter_middleware
>     return await handler(request)
>            ^^^^^^^^^^^^^^^^^^^^^^
>   File "/workspaces/home-assistant-core/homeassistant/components/http/forwarded.py", line 87, in forwarded_middleware
>     return await handler(request)
>            ^^^^^^^^^^^^^^^^^^^^^^
>   File "/workspaces/home-assistant-core/homeassistant/components/http/request_context.py", line 26, in request_context_middleware
>     return await handler(request)
>            ^^^^^^^^^^^^^^^^^^^^^^
>   File "/workspaces/home-assistant-core/homeassistant/components/http/ban.py", line 86, in ban_middleware
>     return await handler(request)
>            ^^^^^^^^^^^^^^^^^^^^^^
>   File "/workspaces/home-assistant-core/homeassistant/components/http/auth.py", line 242, in auth_middleware
>     return await handler(request)
>            ^^^^^^^^^^^^^^^^^^^^^^
>   File "/workspaces/home-assistant-core/homeassistant/components/http/headers.py", line 41, in headers_middleware
>     response = await handler(request)
>                ^^^^^^^^^^^^^^^^^^^^^^
>   File "/workspaces/home-assistant-core/homeassistant/helpers/http.py", line 73, in handle
>     result = await handler(request, **request.match_info)
>              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/workspaces/home-assistant-core/homeassistant/components/http/decorators.py", line 83, in with_admin
>     return await func(self, request, *args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/workspaces/home-assistant-core/homeassistant/components/http/data_validator.py", line 74, in wrapper
>     return await method(view, request, data, *args, **kwargs)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/workspaces/home-assistant-core/homeassistant/components/config/config_entries.py", line 188, in post
>     return await self._post_impl(request, data)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/workspaces/home-assistant-core/homeassistant/components/config/config_entries.py", line 195, in _post_impl
>     return await super()._post_impl(request, data)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/workspaces/home-assistant-core/homeassistant/helpers/data_entry_flow.py", line 92, in _post_impl
>     result = self._prepare_result_json(result)
>   File "/workspaces/home-assistant-core/homeassistant/components/config/config_entries.py", line 215, in _prepare_result_json
>     return _prepare_config_flow_result_json(result, super()._prepare_result_json)
>   File "/workspaces/home-assistant-core/homeassistant/components/config/config_entries.py", line 154, in _prepare_config_flow_result_json
>     return prepare_result_json(result)
>   File "/workspaces/home-assistant-core/homeassistant/helpers/data_entry_flow.py", line 52, in _prepare_result_json
>     data["data_schema"] = voluptuous_serialize.convert(
>                           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
>         schema, custom_serializer=cv.custom_serializer
>         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>     )
>     ^
>   File "/home/vscode/.local/ha-venv/lib/python3.13/site-packages/voluptuous_serialize/__init__.py", line 59, in convert
>     pval = convert(value, custom_serializer=custom_serializer)
>   File "/home/vscode/.local/ha-venv/lib/python3.13/site-packages/voluptuous_serialize/__init__.py", line 164, in convert
>     raise ValueError(f"Unable to convert schema: {schema}")
> ValueError: Unable to convert schema: <function Url at 0x707ab280fe20>

spiddeer committed:
Fix vol.Url() serialization error in config flow

- Replace vol.Url() with str in schema (vol.Url is not serializable)
- Add manual URL validation with urlparse
- Check for scheme and netloc presence
- Wrap in try-except for malformed URLs

Fixes ValueError: Unable to convert schema reported by @sillymoi

spiddeer committed:
Add diagnostics support and bump version to 0.0.4

AndersMarkoff commented:
Looks good. Will merge PR, and release new HACS version tonight.
Thank you for the addons @spiddeer, really appreciate it!
